tests/dynamic_map/retrieve_all_test.cu

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) 2022-2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <test_utils.hpp>
+
+#include <cuco/dynamic_map.cuh>
+
+#include <cuda/std/functional>
+#include <thrust/device_vector.h>
+#include <thrust/equal.h>
+#include <thrust/iterator/zip_iterator.h>
+#include <thrust/sequence.h>
+#include <thrust/sort.h>
+
+#include <catch2/catch_template_test_macros.hpp>
+
+TEMPLATE_TEST_CASE_SIG("dynamic_map retrieve_all tests",
+                       "",
+                       ((typename Key, typename Value), Key, Value),
+                       (int32_t, int32_t),
+                       (int32_t, int64_t),
+                       (int64_t, int32_t),
+                       (int64_t, int64_t))
+{
+  constexpr std::size_t num_keys = 1'000'000;
+  cuco::dynamic_map<Key, Value> map{num_keys * 2,
+                                    cuco::empty_key<Key>{-1},
+                                    cuco::empty_value<Value>{-1},
+                                    cuco::erased_key<Key>{-2}};
+
+  SECTION("retrieve_all after insert")
+  {
+    thrust::device_vector<Key> d_keys(num_keys);
+    thrust::device_vector<Value> d_values(num_keys);
+
+    thrust::sequence(thrust::device, d_keys.begin(), d_keys.end(), 0);
+    thrust::sequence(thrust::device, d_values.begin(), d_values.end(), 0);
+
+    auto pairs_begin =
+      thrust::make_zip_iterator(thrust::make_tuple(d_keys.begin(), d_values.begin()));
```

Comment:
Using a zip iterator to generate keys has known data alignment issues and can potentially lead to runtime hangs, as noted in https://github.com/NVIDIA/cuCollections/pull/677. Ideally, we should always use a transform iterator instead.